### PR TITLE
Gentlemen, it is with great pleasure to inform you that I fixed Reddit Images

### PR DESCRIPTION
**Pull request by N-Brooks**
I was scared that I was gonna have to force everyone to log in themselves and I hated that idea and didn't understand OAuth so I put this off 500 years. But I figured out that I can register my own app with reddit, encrypt the secret, and all is well! This is now an official reddit app registered with them under my phone number/email so hopefully if there are any issues in the future they'll contact me instead of just breaking it and waiting for someone to notice.

### CODE REVIEW

File: apps/redditimages/reddit_images.star [hunk 1]
```diff
@@ -189,15 +217,40 @@ def setRandomPost(allImagePosts, subname):
...
+    body = dict(
+        grant_type = "client_credentials"
+    )
+    res = http.post(
+        url = "https://www.reddit.com/api/v1/access_token",
+        headers = headers,
+        auth = auth,
+        form_body = body,
+        form_encoding = "application/x-www-form-urlencoded",
+        )
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:250]:
```suggestion
    )
```

File: apps/redditimages/reddit_images.star [hunk 2]
```diff
@@ -189,15 +217,40 @@ def setRandomPost(allImagePosts, subname):
...
+    auth = tuple([
+        "ZHJke6sDiMelW-l2l1qqyQ",
+        "QqYyuCX96jXe0QevYe_D7T_LkTHYcw"
+    ])
+    headers = {
+        "User-Agent": "Tidbyt App: Reddit Image Shuffler",
+        "Accept": "application/json"
+        }
+    body = dict(
+        grant_type = "client_credentials"
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:242]:
```suggestion
        grant_type = "client_credentials",
```

File: apps/redditimages/reddit_images.star [hunk 3]
```diff
@@ -189,15 +217,40 @@ def setRandomPost(allImagePosts, subname):
...
+# https://github.com/reddit-archive/reddit/wiki/OAuth2#application-only-oauth
+def getNewAccessToken():
+    authSecret = secret.decrypt(REDDIT_CLIENT_SECRET) or "thiswillfail"
+    auth = tuple([
+        "ZHJke6sDiMelW-l2l1qqyQ",
+        "QqYyuCX96jXe0QevYe_D7T_LkTHYcw"
+    ])
+    headers = {
+        "User-Agent": "Tidbyt App: Reddit Image Shuffler",
+        "Accept": "application/json"
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:239]:
```suggestion
        "Accept": "application/json",
```

File: apps/redditimages/reddit_images.star [hunk 4]
```diff
@@ -189,15 +217,40 @@ def setRandomPost(allImagePosts, subname):
...
+        }
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:240]:
```suggestion
    }
```

File: apps/redditimages/reddit_images.star [hunk 5]
```diff
@@ -189,15 +217,40 @@ def setRandomPost(allImagePosts, subname):
...
             "id": "00000",
         }
 
+# Get a new reddit access token.
+# https://github.com/reddit-archive/reddit/wiki/OAuth2#application-only-oauth
+def getNewAccessToken():
+    authSecret = secret.decrypt(REDDIT_CLIENT_SECRET) or "thiswillfail"
+    auth = tuple([
+        "ZHJke6sDiMelW-l2l1qqyQ",
+        "QqYyuCX96jXe0QevYe_D7T_LkTHYcw"
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:235]:
```suggestion
        "QqYyuCX96jXe0QevYe_D7T_LkTHYcw",
```

File: apps/redditimages/reddit_images.star [hunk 6]
```diff
@@ -141,9 +146,30 @@ def getPosts(subname):
...
     # In lieu of the cache, pull a new set of posts from the API.
-    apiUrl = "https://www.reddit.com/r/" + subname + "/hot.json?limit=30"
-    rep = http.get(apiUrl, headers = {"User-Agent": "Tidbyt App: Reddit Image Shuffler " + str(getRandomNumber(9999))})
+    apiUrl = "https://oauth.reddit.com/r/" + subname + "/hot.json?limit=30"
+    auth = "Bearer " + accessToken
+    rep = http.get(
+        apiUrl,
+        headers = {
+            "User-Agent": "Tidbyt App: Reddit Image Shuffler",
+            "Authorization": auth
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:170]:
```suggestion
            "Authorization": auth,
```

File: apps/redditimages/reddit_images.star [hunk 7]
```diff
@@ -141,9 +146,30 @@ def getPosts(subname):
...
+            }
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:171]:
```suggestion
        }
```

File: apps/redditimages/reddit_images.star [hunk 8]
```diff
@@ -141,9 +146,30 @@ def getPosts(subname):
...
+        )
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:172]:
```suggestion
    )
```

File: apps/redditimages/reddit_images.star [hunk 9]
```diff
@@ -141,9 +146,30 @@ def getPosts(subname):
 
     print("Cache miss, refreshing posts")
 
+    # Pull access token from cache. If expired, get a new one.
+    accessToken = cache.get("reddit-image-posts-access-token")
+    if accessToken == None:
+        print("Access token expired, getting new one")
+        newTokenRes = getNewAccessToken()
+        if "error" in newTokenRes.keys() or newTokenRes["access_token"] ==  None:
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:154]:
```suggestion
        if "error" in newTokenRes.keys() or newTokenRes["access_token"] == None:
```

File: apps/redditimages/reddit_images.star [hunk 10]
```diff
@@ -141,9 +146,30 @@ def getPosts(subname):
...
+            return handleApiError(newTokenRes)
+        accessToken = newTokenRes["access_token"]
+        # Store in cache again
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:157]:
```suggestion

        # Store in cache again
```

File: apps/redditimages/reddit_images.star [hunk 11]
```diff
@@ -141,9 +146,31 @@ def getPosts(subname):
...
-    apiUrl = "https://www.reddit.com/r/" + subname + "/hot.json?limit=30"
-    rep = http.get(apiUrl, headers = {"User-Agent": "Tidbyt App: Reddit Image Shuffler " + str(getRandomNumber(9999))})
+    apiUrl = "https://oauth.reddit.com/r/" + subname + "/hot.json?limit=30"
+    auth = "Bearer " + accessToken
+    rep = http.get(
+        apiUrl,
+        headers = {
+            "User-Agent": "Tidbyt App: Reddit Image Shuffler",
+            "Authorization": auth,
+        }
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:172]:
```suggestion
        },
```

File: apps/redditimages/reddit_images.star [hunk 12]
```diff
@@ -11,7 +11,8 @@ load("encoding/json.star", "json")
 load("http.star", "http")
 load("render.star", "render")
 load("schema.star", "schema")
-load("time.star", "time")
+load("secret.star", "secret")
+load("cache.star", "cache")
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:15]:
```suggestion
```

File: apps/redditimages/reddit_images.star [hunk 13]
```diff
@@ -11,7 +11,7 @@ load("encoding/json.star", "json")
 load("http.star", "http")
 load("render.star", "render")
 load("schema.star", "schema")
-load("time.star", "time")
+load("secret.star", "secret")
```
Comment (matslina) [inline apps/redditimages/reddit_images.star:14]:
```suggestion
load("secret.star", "secret")
load("time.star", "time")
```

**Comment by matslina** [comment]
@N-Brooks I fixed up most (hopefully all) of the `pixlet check` errors. You're left with lint errors now. An unused variable and an unreachable statement.

**Comment by N-Brooks** [comment]
Sorry about the delay. I realized that I accidentally committed the secret I wanted to use by that lint so I wiped it out and replaced it with a new one.

**Pull request merged**